Add Linux support policy (#4823)
jenkins-infra/jenkins.io: content/doc/administration/requirements/linux.adoc

---
layout: documentation
title:  Linux Support Policy
---

This page documents the Linux support policy for the Jenkins controller and agents.

== Scope

> Jenkins plugins may set additional requirements to Linux versions on controllers and/or agents.

Reviewer: ```suggestion
Individual Jenkins plugins may set additional requirements for Linux versions on controllers and/or agents.
```

This page does not document such requirements.
Please refer to plugin documentation.

== Why?

Theoretically, Jenkins can run everywhere where you can run a supported Java version,
but there are some limitations in practice.
Jenkins core and some plugins include native code or depend on Linux API and subsystems,
and hence they rely on specific Linux versions.
Jenkins platform specific installation packages rely on specific Linux versions.

== Support levels

We define multiple support levels for Linux platforms.

[width="100%",cols="20%,35%,45%",options="header",]
|===
|Support level |Description |Platforms

| **Level 1** - Full support

> | We run automated testing for these platforms, and we intend to fix the reported issues timely.

Reviewer: ```suggestion
| We run automated testing for these platforms, and we intend to fix the reported issues in a timely manner.
```

a|

>   * 64-bit (amd64) Linux versions supported by Debian, Red Hat, SUSE, and Ubuntu

Reviewer 2: Do we really run automated tests for all these platforms, I do not think so

Author: Good point.  We don't run automated tests on all those platforms.  We run automated tests of Jenkins core builds on Ubuntu and Windows Server 2019.  We run automated tests of weekly installers on Debian and CentOS.  We run automated tests of LTS installers on Debian and CentOS.

Should this draft and the [Windows Support Policy](https://www.jenkins.io/doc/administration/requirements/windows/) be modified to more precisely state what we test with automation?

Alternately, the statements about automated tests could be removed, since the users probably care more that we are actively maintaining certain platforms rather than the specific testing methods we use.

Author: Thanks to the work from @basil in https://github.com/jenkinsci/packaging, we now have [package installation tests](https://ci.jenkins.io/job/Packaging/job/packaging/job/master/) that check:

* almalinux-8
* amazonlinux-2
* centos-stream-8
* debian-10
* debian-11
* fedora-34
* fedora-35
* opensuse-leap-15
* oraclelinux-8
* rockylinux-8
* ubuntu-18-04
* ubuntu-20-04
* ubuntu-21-04
* ubuntu-21-10
* ubuntu-22-04

  * Linux versions used in the official Docker images

| **Level 2** - Supported

> | We do not actively test these platforms, but we intend to keep compatibility.

Reviewer 3: > We do not actively test these platforms, but we intend to keep compatibility.

How do we intend to keep compatibility when making large-scale changes if we do not actively test these platforms?

Author: > How do we intend to keep compatibility when making large-scale changes if we do not actively test these platforms?

We rely on users and contributors to test these platforms.  When they encounter an issue, they report it.

In the case of some of the platforms in the list (like arm64 and s390x on Debian), we are actively testing them with the [weekly installation acceptance test job](https://ci.jenkins.io/job/Infra/job/acceptance-tests/job/install-latest-debian-package/) and the [LTS installation acceptance test job](https://ci.jenkins.io/job/Infra/job/acceptance-tests/job/install-lts-debian-package/).  However, that little bit of installation testing was not enough for me to feel that we should list those platforms as "Level 1".

Reviewer 3: > We rely on users contributors to test these platforms

So a contributor making a large-scale change would be expected to test that large-scale change on Level 2 platforms (or solicit testing by users and debug by telegram) prior to integration, correct? That does not seem congruent with the notion of "we do not actively test these platforms". And who is the "we" referring to, anyway? Maintainers? Are maintainers not contributors, too?

Author: > > We rely on users and contributors to test these platforms
> 
> So a contributor making a large-scale change would be expected to test that large-scale change on Level 2 platforms (or solicit testing by users and debug by telegram) prior to integration, correct? That does not seem congruent with the notion of "we do not actively test these platforms". And who is the "we" referring to, anyway? Maintainers? Are maintainers not contributors, too?

I should have been more precise.  I should have said.

> We rely on users to test these platforms

I was using an overly broad definition of "contributor" that is indistinguishable from "user".  Should have kept the phrasing simple and said "users" instead of "users and contributors".

Reviewer 3: Got it. Then a contributor (who may or may not also be a maintainer) making a large-scale change would have to rely on users to test the change on Level 2 platforms (and possibly debug issues by telegram) before the change could be integrated. This seems undesirable to me.

Author: > Then a contributor (who may or may not also be a maintainer) making a large-scale change would have to rely on users to test the change on Level 2 platforms (and possibly debug issues by telegram) before the change could be integrated. This seems undesirable to me.

The policy does not limit what a contributor can do.  It tries to describe the minimum levels we are striving to do.

A contributor is welcome to do more testing before the change is integrated.

Reviewer 3: Of course, but if a contributor is _required_ to either do that testing or to solicit users to do it on the contributor's behalf, then IMHO the platform is "fully" supported.

  We will consider patches if they do not put Level 1 support at risk and do not create maintenance overhead.
a|
  * 64-bit (amd64, arm64, s390x, ppc64le) Linux versions that use the Debian packaging format

>   * 64-bit (amd64, arm64, s390x, ppc64le) Linux versions that use the rpm packaging format

Reviewer 2: Packaging format has nothing to do with the supported platforms. What is the intent of excluding other platforms?

Author: I hadn't considered a user that chooses to use the war file rather than using a package manager.  I'm open to suggestions of alternate phrasing that would also cover users running with a war file.

Author: I've included phrasing for war file users and for container users.

| **Level 3** - Patches considered
| Support may have limitations and extra requirements.

>   We do not test compatibility, and we may drop support if there is a need.

Reviewer: ```suggestion
  We do not test compatibility, and we may drop support at any time.
```
I thought about "we may drop support when necessary" but isn't the point that we can drop support at any time?

>   We will consider patches if they do not put level 1/2 support at risk and do not create maintenance overhead.

Reviewer: ```suggestion
  We consider patches that do not put level 1/2 support at risk and do not create maintenance overhead.
```

a|
  * 32-bit (x86, arm) Linux versions
  * RISC-V and other architectures not included in level 1 or level 2 support
  * Preview releases

| **Level 4** - Unsupported
| These versions are known to be incompatible or to have severe limitations.

>   We do not support the listed platforms, and we will not accept patches.

Reviewer: ```suggestion
  We do not support the listed platforms, and we do not accept patches.
```

a|

>   * Linux versions no longer supported by operating system providers

Reviewer 2: I do not think it is correct to make it unsupported, because it is common to see version outside of the EoL being used. Also I am not sure who is "operating system providers" for Linux distributions that can be built locally

Author: > I do not think it is correct to make it unsupported, because it is common to see version outside of the EoL being used. 

I have been unwilling to spend any effort adapting to Linux versions that are no longer supported by the operating system provider.  I've not seen anyone else willing to spend effort adapting to Linux versions that are no longer supported by their provider.

> Also I am not sure who is "operating system providers" for Linux distributions that can be built locally

If there is no end of life for an operating system, then the phrase does not apply.

|===

== References

* link:https://wiki.debian.org/LTS[Debian Long Term Support]
* link:https://access.redhat.com/support/policy/updates/errata[Red Hat Enterprise Linux Life Cycle]
* link:https://www.suse.com/lifecycle/[SUSE Product Support Lifecycle]
* link:https://ubuntu.com/about/release-cycle[Ubuntu lifecycle and release cadence]

== Contributing

> If you would like to add support for more Linux platforms or to share feedback,
> we will appreciate your contributions!

Reviewer: ```suggestion
You are welcome to propose PR's that add support for other Linux platforms or to share feedback;
we will appreciate your contributions!
```
I'm not sure if PR's is the right noun to use here.  Packages, software, ...?  But just "add support" seems ambiguous -- anyone can add support to their Jenkins installation.  I think we are talking about what they contribute back to the produce, and there they can propose software that the community does not accept.
